bench/src/main/scala/cats/bench/UnorderedTraverseMapBench.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2015 Typelevel
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package cats.bench
+
+import org.openjdk.jmh.annotations._
+import cats.{CommutativeApplicative, Eval, UnorderedTraverse}
+import java.util.concurrent.TimeUnit
+import cats.data.Chain
+
+@State(Scope.Benchmark)
+@BenchmarkMode(Array(Mode.AverageTime))
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+class UnorderedTraverseMapBench {
+  // These benchmarks were written to choose the fastest implementation
+  // for Map.unorderedTraverse, some results are available here:
+  // https://github.com/typelevel/cats/pull/4463#issuecomment-1599612154
+
+  val instance = UnorderedTraverse[Map[Int, *]]
+
+  val xs1: Map[Int, Int] = (1 to 1000).map(x => (x, x)).toMap
+  val xs2: Map[Int, Int] = (1 to 1000000).map(x => (x, x)).toMap
+
+  def unorderedTraverseViaTree[G[_], A, B](
+    fa: Map[Int, A]
+  )(f: A => G[B])(implicit G: CommutativeApplicative[G]): G[Map[Int, B]] = {
+    def runHalf(size: Int, fa: Map[Int, A]): Eval[G[Map[Int, B]]] =
+      if (size > 1) {
+        val leftSize = size / 2
+        val rightSize = size - leftSize
+        val (leftL, rightL) = fa.splitAt(leftSize)
+        runHalf(leftSize, leftL).flatMap { left =>
+          val right = runHalf(rightSize, rightL)
```

Comment:
since we have this I do wonder how much it costs to run `Eval.defer(runHalf(rightSize, rightL))` here which would allow short circuiting for short circuiting Applicatives (Option, Either, Try, IO).

Currently we do walk the entire structure.

Reply:
Hmm, that might be worth a bench too

Reply:
I was thinking about it, and probably we use these as seeds to benchmark your [Traverse.Strategy](https://github.com/typelevel/cats/issues/3790#issuecomment-896267508) (ofc we'll need to implement it).
Also, I've taken a look at [Daniel's benchmarks](https://github.com/typelevel/cats/pull/4403) and it seems that we need to refactor them to avoid inlining and other stuff. In any case, I'll need to study a bit how to better use jmh in order to write better benchmarks.